### ISSUE TITLE
GML writer can't handle quotes

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#2011]
**Describe the bug**
When exporting string attributes whose value has a `"`, the GML writer produces an invalid output. It does not quote the strings.

**To reproduce**
Export a graph with a string attribute value such as `"a\"bc"`. The result will be written as `"a"bc"`.

**Version information**
Current.

**Comment by szhorvat** [comment]
The GML spec has to say this:

> Further reglementations are a maximum line length and a maximum key size of 254 characters (this is neccessary since some operating systems and editors do not handle longer lines), and the use of 7-bit ASCII characters only. Any other characters are coded in the ISO 8859-1 character set, and have the form &name;. Especially, the characters " and & within strings must be coded this way to avoid ambiguity. The ISO 8859-1 ist also used by HTML, which is the most common format for distributing data on the world wide web.

Thus, technically both `&` and `"` are forbidden. Do other GML implementations follow this? I'm not sure—the yEd writer seems to.

The compromise I suggest is to handle `"` (and only `"`) in the writer, but not the reader, and convert it to `&quot;`. This will at least produce a GML file that most other software can import.

**Comment by szhorvat** [comment]
One of the reasons why I was looking at quotes is that currently igraph cannot import "composite" GML attributes because it cannot store composite attributes, except in a host-language specific way. For example, some software might export:

```
node [
  id 1
  coord [
    x 0 y 1
  ]
]
```

In fact, many will export something like this:

```
node [
  id 1
  graphics [
    x 0.0
    y 0.0
  ]
```

Thus igraph will not be able to read the coordinates.

How could we get this data into Python, R or Mathematica? We'd need some format to transmit in. One idea that is relatively easily workable is to convert composites to JSON and store them as strings. Most host languages have facilities to deserialize JSON. But JSON contains a lot of quotes, so this way we end up with attribute values that have quotes, which would cause trouble on re-exporting.

I'm not sure I want to do the JSON idea now, but handling quotes should be fixed soon.

**Comment by ntamas** [comment]
> The compromise I suggest is to handle " (and only ") in the writer, but not the reader, and convert it to &quot;. This will at least produce a GML file that most other software can import.

I think we should handle at least `"` and `&`, and do it both in the reader and the writer so we can make a round-trip conversion from igraph to GML to igraph without loss of information if the original graph contains only ISO-8859-1 data. This can get into 0.9.x or 0.10.x as it seems to be a minor change.

Encodings are going to be a bit tricky. Right now there is no indication of the encoding of string attributes in igraph. The _de facto_ standard has been UTF-8 in the Python interface; I don't know how other inferfaces handle this, though. GML is not able to handle all UTF-8 characters, so we would need to recode strings from UTF-8 to ISO-8859-1 if we wanted to do this right before escaping them with ampersands. This is actually a lucky combination because ISO-8859-1 is a subset of Unicode (so we can get away without using `iconv` or some other library to get the conversion right), but still I'd leave this for some version later than 0.10.

**Comment by szhorvat** [comment]
I have some concerns about trying to do a too-complicated encoding/decoding. I put this on the meeting agenda not this week (I can't come) but in two weeks. It's easier to discuss in person than in writing.

This will definitely not go into 0.9. There's too much refactoring in the GML reader since then. It must go in 0.10. Although the encoding/decoding step should be properly separated from the GML parsing, so that it can be worked on independently. Maybe we can leverage libxml2 for this if we want more than `&quot;` and `&amp;`?

**Comment by szhorvat** [comment]
In the meantime if anyone wants to do a survey of other libraries and see how they handle this, that will be very useful. We want to go for the de-facto format instead of sticking to an ancient specification which I had to dig out from the Internet Archive. List of software supporting GML is here: https://en.wikipedia.org/wiki/Graph_Modelling_Language

**Comment by ntamas** [comment]
> Maybe we can leverage libxml2 for this if we want more than &quot; and &amp;?

I don't think we need `libxml2` for that. In theory, ampersand-encoding a character with code `x` the way it is done in XML files is simply equal to `&#x;`, i.e. `"` is `&#34;`, `&` is `&#38;` and so on. `&quot;` and `&amp;` are just convenience aliases for readability, so we don't need to rely on `libxml2` to get access to a table of commonly used aliases (like `&lt;` and `&gt;` for `<` and `>`) because they are all valid.

_Edit_: actually, you are right, we should take a look at what other software do when writing or reading GML files. Many of the parsers out there seem to be broken. For instance, [this parser](https://github.com/marcusraitner/gml_parser/blob/master/gml_scanner.c#L17) does not know about the `&#number;` syntax -- it contains a hardcoded encoding for all characters in ISO-8859-1 that are not ASCII. The problem is that in XML, only `&lt;`, `&gt;`, `&quot;`, `&amp;` and `&apos;` exist implicitly ([reference](https://www.w3.org/TR/2008/REC-xml-20081126/#sec-predefined-ent)). The entities in that parser are not from XML, they are from HTML, which is a superset of XML. I don't think that the GML specification mentions HTML anywhere - but if we used `libxml2` for ampersand-encoding, we would not get the HTML entities by default as `libxml2` knows that XML != HTML.

**Comment by szhorvat** [comment]
Implementing this properly is a lot of work, and I do not expect it to be done for 0.10. I would like to have a stopgap measure in the meantime. **I propose skipping all attributes with an invalid name, with a clear warning.** Then users can manually fix attribute names before export. This is better than writing a corrupt file that cannot be read back.

Shall we go ahead with this?

**Comment by ntamas** [comment]
Seems good to me, but let's leave the issue open after that. I think we should eventually implement the handling of `&quot;` and `&amp;` "manually", preferably both in the reader and the writer (so we can read back what we write), as these seem to be used by other tools that support GML. Everything else would probably be too much work for too little gain, considering that there is no formal specification that would explicitly lay out what is allowed and what is not.

**Comment by szhorvat** [comment]
@ntamas Regarding the handling of nested properties such as:

```
node [
  id 1
  graphics [
    x 0.0
    y 0.0
  ]
```

One option would be to encode these as two node attributes called `graphics.x` and `graphics.y` in igraph. This is easy to implement in the reader, and it is useful, as `graphics` is a de-facto standard part of GML. However, it is not so easy to implement in the writer. 

`.` cannot be a part of GML keys, so there is no chance for conflict.

**Comment by szhorvat** [comment]
Fixed by #2059.
